### PR TITLE
Make Stream.withFilter.{map,flatMap} run in constant stack space

### PR DESCRIPTION
**Pull request by Blaisorblade**
The included test currently fails because `map` and `flatMap` do not
run in constant stack space on a stream returned by `Stream.withFilter`,
as I reported here:
https://groups.google.com/d/msg/scala-language/WqJR38REXnk/saaSiDdmyqoJ

Fix the problem and add a simple testcase.

Note that the stack space consumed when producing an element of this stream is
proportional to the number of elements failing the test before the next
success. The stack space consumed to produce the stream itself is the
space needed to produce the first element, that is, is proportional to
the number of failures before the first success.

Review by @axel22.

### CODE REVIEW

File: test/files/run/stream-stack-overflow-filter-map.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+object Test extends App {
+  //This runs fine.
+  val resFMap1 = (1 to 10000).toStream filter (_ => false) flatMap (Seq(_))
+  val resMap1 = (1 to 10000).toStream filter (_ => false) map (_ + 1)
+  assert(resMap1.isEmpty)
+  assert(resFMap1.isEmpty)
+  println(resMap1)
+  println(resFMap1)
```
Comment (Blaisorblade) [inline test/files/run/stream-stack-overflow-filter-map.scala]: On second thought, I could remove these lines and have no output at all. I'll push a new commit if this is preferred.

File: test/files/run/stream-stack-overflow-filter-map.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  //This will cause a stack overflow
```
Comment (Blaisorblade) [inline test/files/run/stream-stack-overflow-filter-map.scala]: Here I meant "with the previous implementation".

File: src/library/scala/collection/immutable/Stream.scala [hunk 3]
```diff
@@ -479,22 +479,57 @@ self =>
...
+
+      /*
       def tailMap = asStream[B](tail withFilter p map f)
       if (isStreamBuilder(bf)) asThat(
         if (isEmpty) Stream.Empty
         else if (p(head)) cons(f(head), tailMap)
         else tailMap
+        //XXX Alternative: what about having a Stream.step constructor?
       )
+      */
```
Comment (Blaisorblade) [inline src/library/scala/collection/immutable/Stream.scala]: Commenting out code instead of removing it was not my intention. Will fix.

**Comment by Blaisorblade** [comment]
PLS REBUILD ALL

**Comment by Blaisorblade** [comment]
Note: the test results are, strictly speaking, not valid any more, even though the changes are minor. I would have deleted the comments, as required by the pull request policy, but I don't have permissions for that.

**Comment by axel22** [comment]
I believe this would benefit from additional correctness test cases such as calling `withFilter` and `map`/`flatMap` on the:
- empty stream
- one-element stream
- two-element stream

In all cases, the predicate for `withFilter` should a) always return true, b) always return false, c) `_ % 2 == 0`, d) `_ % 3 == 0`.
In all cases, the result should be compared against the expected result to see if the correct stream elements are returned.

After that, LGTM.

**Comment by Blaisorblade** [comment]
I wrote the test case you asked for - however, the "expected" result is computed by converting the stream to a Seq (and checking that the result is right) and using filter, map and flatMap on Seq. I thus assume that those functions on Seq are are tested independently. If that's a problem I could list instead the results explicitly, even though that would make additional testcases harder to add. Or I could additionally print the results (and have a corresponding .check file).

**Comment by Blaisorblade** [comment]
PLS REBUILD ALL

**Comment by axel22** [comment]
LGTM

**Pull request merged**